### ISSUE TITLE
Add a types header

### ISSUE THREAD

**Issue by rofinn** [JuliaData/CSV.jl#380]
It'd be nice if you could tell `CSV` to write a types header line that would make it easier to read back into julia.

```
a,b,c,d
DateTime,String,Float64,MyType
...
```

Reading a value back could just always be `parse(eval(Symbol(col_type)), x)`. The `eval` isn't great, but it's restrict to only symbols. 

This approach would put the onus on package developers to appropriately define `parse` methods for their types, rather than the person loading the CSV.

**Comment by quinnj** [comment]
Hmmmm.....definitely an interesting idea. Here's a couple thoughts off the top of my head:
* How would reading it back in work exactly? Do we create a separate `CSV.jlread` function that expects the 2nd row to have types? Does `CSV.read` somehow check that all the comma separated values in the 2nd row are valid types? Do we add an extra keyword argument to `CSV.read` to expect and use the 2nd row of types?
* We could have a `Dict{String, Type}` lookup for the most common types and only use `eval` when a custom type is encountered
* For columns with `missing` values, I assume we'd write `Union{Missing, T}`? That should be fine I guess, but we might have to put it in `"` so the extra commas don't throw the parser off
* While we're writing out columns types, we could potentially also write out the number of rows. That would speed up reading a bit as well. Not sure if that would be done as a 3rd row, for each column, or just as a single number or what.
* If you wanted to read the file from a different language, I guess you would just make sure to specify to skip the 2nd row when parsing (most parsers support some kind of option like this)

Anyway, I don't see any reason we couldn't do something like this.

**Comment by nalimilan** [comment]
This should probably be in a comment so that it doesn't disturb other parsers. It could start with a special sequence like `# Types:` or `# Julia Types:`, so that automatic detection will reliably work with a very low probability of false positives.

Note that there already exist specifications which address this goal. One of them is [linked CSV](http://jenit.github.io/linked-csv/#prolog-lines), but their approach requires adding a column, which changes the parsing by other implementations. [RGDAL](https://www.gdal.org/drv_csv.html) uses a separate .csvt file for this.

**Comment by rofinn** [comment]
> Do we add an extra keyword argument to CSV.read to expect and use the 2nd row of types?

Yes, I was thinking that both `CSV.read` and `CSV.write` would have the same keyword argument to include the types header, so that it's explicit when calling.

> We could have a Dict{String, Type} lookup for the most common types and only use eval when a custom type is encountered

Can you clarify what the advantage of that would be? Is `eval` particularly slow? Would you want to throw a warning when needing call `parse(eval(Symbol(col_type)), x)`? Would you want this to be a way to infer whether the second line is a types header, so the keyword argument isn't always required?

> This should probably be in a comment so that it doesn't disturb other parsers. It could start with a special sequence like # Types: or # Julia Types:, so that automatic detection will reliably work with a very low probability of false positives.

The only issue is that comments aren't officially part of the CSV spec

**Comment by quinnj** [comment]
Ok, I'm open to this if someone wants to take a stab at it. Essentially the work to do would be:

* in `CSV.write` for *known schema* (there are 2 main `write` functions, one for known input schema, one for unknown), we would add a keyword argument `types::Bool=false`, which, if `true`, would write out a row of column types after the header row
* if the column type is `Union{T, Missing}`, we'd need to write it out as `\"Union{T, Missing}\"` (i.e. quoted using `openquotechar` and `closequotechar`)
* for `CSV.File`/`CSV.read`, we could allow the `types` keyword argument to be an `Integer` for the "types row" number (or perhaps just a Bool, so if `true`, it would mean right after the `header` row). We would pass this arg to the `datalayout` functions and based on where/how we get the column names, we'd then read the types (if applicable), then move on to the `datarow`.
* So `datalayout` would return an additional `parsedtypes` return value, which would be used to set the `typecodes` (if not empty obviously)

**Comment by Moelf** [comment]
while this may be good for some cases, I personally don't like the idea of 'breaking' a CSV file; sure, you can skip in some libraries, but it's still a 'bad' CSV file we're producing, much like the ones that have a trailing line of meta data, [example](http://ftp.nasdaqtrader.com/dynamic/SymDir/otherlisted.txt).

**Comment by rofinn** [comment]
How is it a 'bad' CSV file? AFAIK, nothing in the spec prevents a CSV file from having extra metadata lines. Most CSV parsers that I know of allow you to skip/declare header lines and it'd be easy enough to manual parse header and data lines independently if necessary.

**Comment by Moelf** [comment]
if you have a second after the header, it will be parsed into the data in most default cases.

**Comment by rofinn** [comment]
True, but that doesn't make it a 'bad' CSV file... it just means that the content **can** be julia specific. Again, the goal isn't to have this behaviour by default, but it would make it easier for folks who want to save/load their tabular data in a human readable form.
